File: binarytree/101_is_symmertic.py
from typing import List, Deque
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:

    def dfs(self, a: TreeNode, b: TreeNode) -> bool:
        if not (a and b):
            return a is None and b is None
        else:
            return a.val == b.val and self.dfs(a.left, b.right) and self.dfs(a.right, b.left)
        

    def isSymmetric(self, root: TreeNode) -> bool:
        return self.dfs(root.left, root.right)


if __name__ == '__main__':
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    print(Solution().isSymmetric(root))

File: binarytree/101_is_symmertic.cc
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    bool dfs(TreeNode *a, TreeNode *b)
    {
        if (a == nullptr && b == nullptr)
        {
            return true;
        }
        if (a == nullptr || b == nullptr)
        {
            return false;
        }
        return a->val == b->val && dfs(a->left, b->right) && dfs(a->right, b->left);
    }

    bool isSymmetric(TreeNode *root)
    {
        return dfs(root, root);
    }
};